# Análisis de datos con Pyhon - Pre-entrega
Nombre y apellido:

# Etapa 1

## Script básico
Actividades 2 y 3

In [ ]:
# Generamos una lista de diccionarios que contenga algunos datos
ventas = [
  {'producto': 'notebook', 'precio': 1500.0, 'cantidad': 3},
  {'producto': 'monitor', 'precio': 500.0, 'cantidad': 5},
  {'producto': 'mouse', 'precio': 20.0, 'cantidad': 15}
]

In [ ]:
# Calculamos las ventas acumuladas - map


In [ ]:
# Calculamos las ventas acumuladas - listas por comprensión
venta_acumuladas = sum([venta["precio"] * venta["cantidad"] for venta in ventas])
venta_acumuladas

## Import Datasets
Actividad 1

In [2]:
import pandas as pd

In [3]:
# Import dataset ventas
df_ventas = pd.read_csv("https://raw.githubusercontent.com/jrigottiedu/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/PreEntrega/Datasets/ventas.csv")

In [4]:
# Import dataset clientes
df_clientes = pd.read_csv("https://raw.githubusercontent.com/jrigottiedu/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/PreEntrega/Datasets/clientes.csv")

In [5]:
# Import dataset marketing
df_marketing = pd.read_csv("https://raw.githubusercontent.com/jrigottiedu/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/PreEntrega/Datasets/marketing.csv")

In [6]:
df_ventas

id_venta                producto   precio  cantidad fecha_venta  \
0          792       Cuadro decorativo   $69.94       5.0  02/01/2024   
1          811         Lámpara de mesa  $105.10       5.0  02/01/2024   
2         1156                Secadora   $97.96       3.0  02/01/2024   
3         1372                Heladera  $114.35       8.0  02/01/2024   
4         1546                Secadora  $106.21       4.0  02/01/2024   
...        ...                     ...      ...       ...         ...   
3030      1837         Horno eléctrico  $104.12       9.0  30/12/2024   
3031      2276                  Laptop   $85.27       9.0  30/12/2024   
3032      2696                  Laptop  $107.81       4.0  30/12/2024   
3033      2913              Smartphone   $99.85       7.0  30/12/2024   
3034      2930  Consola de videojuegos   $55.47       6.0  30/12/2024   

              categoria  
0            Decoración  
1            Decoración  
2     Electrodomésticos  
3     Electrodomésticos  
4     Electrodomésticos  
...                 ...  
3030  Electrodomésticos  
3031        Electrónica  
3032        Electrónica  
3033        Electrónica  
3034        Electrónica  

[3035 rows x 6 columns]

## Dataframe ventas

### Análisis exploratorio de **datos**
Actividad 4

Aplicamos los métodos de EDA vistos en clase, `info`, `columns`, `dtypes`, `head`, `tail`, `sample`, `unique`, `value_counts`, `describe`, `duplicated`, `isnull`, etc. y documentamos las observaciones:

In [7]:
# Aplicamos info() para familiarizarnos con el dataframe
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3035 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     3035 non-null   int64  
 1   producto     3035 non-null   object 
 2   precio       3033 non-null   object 
 3   cantidad     3033 non-null   float64
 4   fecha_venta  3035 non-null   object 
 5   categoria    3035 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 142.4+ KB


In [ ]:
# Visualizamos los primeros registros con head()
# también podemos usar tail() para visualizar los últimos
df_ventas.head()

In [ ]:
# Usar sample() para tomar registros al azar
df_ventas.sample(5)

In [ ]:
# Analizamos el campo precio
# Si es numericos se puede aplicar describe()
df_ventas["precio"].sample(5)

In [ ]:
# Analizamos el campo cantidad
# si es numerico se puede aplicar describe()
df_ventas["cantidad"].sample(5)

In [ ]:
# Analizamos el campo fecha_venta
df_ventas["fecha_venta"].sample(5)

In [ ]:
# Analizamos el campo categoria
# Probar con value_counts y unique
df_ventas["categoria"].value_counts()

### Calidad de datos
Actividad 5

In [19]:
# Buscamos registros duplicados
df_ventas.duplicated().sum()

np.int64(35)

In [20]:
# Visualizamos los duplicados
# keep: 'first', 'last', False
df_ventas[df_ventas.duplicated(keep=False)].sort_values(by="producto")

id_venta         producto   precio  cantidad fecha_venta    categoria
845       300  Adorno de pared   $67.75       9.0  06/04/2024   Decoración
833       300  Adorno de pared   $67.75       9.0  06/04/2024   Decoración
857       345         Alfombra   $56.78       9.0  07/04/2024   Decoración
832       145         Alfombra   $41.72       8.0  06/04/2024   Decoración
868       345         Alfombra   $56.78       9.0  07/04/2024   Decoración
..        ...              ...      ...       ...         ...          ...
854      2108       SmartWatch   $79.37      10.0  06/04/2024  Electrónica
831      2778       SmartWatch   $70.66       7.0  05/04/2024  Electrónica
825      2778       SmartWatch   $70.66       7.0  05/04/2024  Electrónica
877      2373       Smartphone  $123.60      12.0  07/04/2024  Electrónica
866      2373       Smartphone  $123.60      12.0  07/04/2024  Electrónica

[70 rows x 6 columns]

In [21]:
# Buscamos datos nulos
df_ventas.isnull().sum()

id_venta       0
producto       0
precio         2
cantidad       2
fecha_venta    0
categoria      0
dtype: int64

In [22]:
# Visualizar los nulos
# df_ventas[df_ventas.isnull().any(axis=1)]
df_ventas[df_ventas.isnull().any(axis=1)]

id_venta               producto precio  cantidad fecha_venta    categoria
136       627  Elementos de cerámica    NaN       NaN  17/01/2024   Decoración
139      2171    Parlantes Bluetooth    NaN       NaN  17/01/2024  Electrónica

In [23]:
# Analizar los precios de los NaN a ver que opciones tenemos
# df_ventas[df_ventas["producto"]=="Elementos de cerámica"]
# df_ventas["producto"].unique()
# agrupo por productos iguales y suma el total
df_ventas["producto"].value_counts()


producto
Lámpara de mesa           181
Auriculares               147
Microondas                135
Cafetera                  118
Laptop                    104
Secadora                  103
Jarrón decorativo         102
Smartphone                102
Alfombra                  102
SmartWatch                102
Aspiradora                102
Cortinas                  102
Rincón de plantas         102
Freidora eléctrica        101
Espejo decorativo         101
Adorno de pared           101
Lavadora                  100
Consola de videojuegos    100
Cuadro decorativo         100
Horno eléctrico           100
Plancha de vapor          100
Proyector                 100
Elementos de cerámica     100
Batidora                  100
Televisor                 100
Cámara digital            100
Parlantes Bluetooth        99
Tablet                     58
Heladera                   49
Candelabro                 24
Name: count, dtype: int64

### Observaciones




* Cantidad de registros: 3035
* Columnas 6
* 35 duplicados, eliminarlos
* Algunos nulos (2), 1% del faltante, dropna
* Campo precio: encontre `$`, removerlo y luego cambiar el dtype a float64
* Campo fecha: ajustar a dtype datetime

## Dataframe marketing

### Analisis exploratorio de datos

Aplicamos los métodos de EDA vistos en clase, `info`, `columns`, `dtypes`, `head`, `tail`, `sample`, `unique`, `value_counts`, `describe`, `duplicated`, `isnull`, etc. y documentamos las observaciones:

In [ ]:
# Aplicamos info() para familiarizarnos con el dataframe
df_marketing.info()

In [9]:
# Visualizamos los primeros registros con head()
# también podemos usar tail() para visualizar los últimos
# df_marketing.tail()
df_marketing.head()

id_campanha         producto  canal  costo fecha_inicio   fecha_fin
0           74  Adorno de pared     TV   4.81   20/03/2024  03/05/2024
1           12           Tablet   RRSS   3.40   26/03/2024  13/05/2024
2           32  Lámpara de mesa  Email   5.54   28/03/2024  20/04/2024
3           21       Smartphone   RRSS   6.37   29/03/2024  16/05/2024
4           58         Alfombra  Email   4.25   31/03/2024  05/05/2024

In [14]:
# Analizamos el campo costo
df_marketing["costo"].sample(5)

77    4.10
49    3.56
80    5.62
50    5.28
46    4.94
Name: costo, dtype: float64

In [15]:
# Analizamos el campo fecha_inicio
df_marketing["fecha_inicio"].sample(5)

61    22/09/2024
77    03/11/2024
21    30/05/2024
10    27/04/2024
37    16/07/2024
Name: fecha_inicio, dtype: object

In [16]:
# Analizamos el campo fecha_fin
df_marketing["fecha_fin"].sample(5)

64    15/11/2024
70    21/12/2024
35    15/08/2024
36    03/08/2024
78    23/12/2024
Name: fecha_fin, dtype: object

### Calidad de los datos


In [18]:
# Buscamos registros duplicados
df_marketing.duplicated().sum()

np.int64(0)

### Observaciones
*** Completar ****

## Dataframe clientes

### Analisis exploratorio de datos

### Observaciones

# Etapa 2

## Limpieza de datos
Actividad 1 - Limpiar el conjunto de datos eliminando duplicados y caracteres no deseados. Documentar el proceso y los resultados.

### Dataframe ventas  (acciones)

* 35 duplicados, eliminarlos - se eliminaron 35 duplicados
* Algunos nulos (2), 1% del faltante, dropna - se eliminaron 2 registros con 2 NaN
* Campo precio: encontre `$`, removerlo y luego cambiar el dtype a float64 - Ok
* Campo fecha: ajustar a dtype datetime - Ok

In [ ]:
# Eliminamos registros duplicados
# atributos a evaluar, keep, ignore_index, inplace
df_ventas.drop_duplicates(inplace=True, ignore_index=True)

In [ ]:
# Validamos que no haya duplicados
df_ventas.duplicated().sum()

In [ ]:
# Eliminamos registros con datos nulos
df_ventas.dropna(inplace=True)

In [28]:
# Validamos que no haya nulos
# df_ventas.isnull().sum()
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3035 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     3035 non-null   int64  
 1   producto     3035 non-null   object 
 2   precio       3033 non-null   object 
 3   cantidad     3033 non-null   float64
 4   fecha_venta  3035 non-null   object 
 5   categoria    3035 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 142.4+ KB


In [26]:
# Eliminamos el caracter especial $ de precio
# Normalización: usar replace()
# De ser necesario, podemos usar el atributo regex=True
df_ventas["precio"] = df_ventas["precio"].str.strip().str.replace("$", "")

In [27]:
# Validamos que precio no contenga $
df_ventas["precio"].sample(3)

304      49.23
1047     82.23
36      108.48
Name: precio, dtype: object

In [29]:
# Convertimos precio de str a float
# Podemos usar pd.to_numeric o astype o incluso combinarlos
df_ventas["precio"] = df_ventas["precio"].astype(float)

In [30]:
# Convertir el dtype de cantidad, de float64 a int64
df_ventas["cantidad"]= df_ventas["cantidad"].astype(int)

IntCastingNaNError: Cannot convert non-finite values (NA or inf) to integer

In [ ]:
# Validar la conversion a int
df_ventas["cantidad"].dtypes

In [ ]:
# Convertir el campo fecha_venta a dtype datetime
# pd.to_datetime(df_ventas["fecha_venta"], format="%d/%m/%Y")
# pd.to_datetime(df_ventas["fecha_venta"], dayfirst=True)
df_ventas['fecha_venta'] = pd.to_datetime(df_ventas['fecha_venta'], format="%d/%m/%Y")

In [ ]:
# Validar la conversion de fecha
df_ventas['fecha_venta'].dtype

### Dataframe marketing

Convertir a datetime fecha_inicio / fecha_fin

In [ ]:
df_marketing.info()

In [ ]:
# Convertir el campo fecha_inicio a dtype datetime
df_marketing['fecha_inicio'] = pd.to_datetime(df_marketing['fecha_inicio'], format="%d/%m/%Y")

In [ ]:
# Convertir el campo fecha_fin a dtype datetime
df_marketing['fecha_fin'] = pd.to_datetime(df_marketing['fecha_fin'], format="%d/%m/%Y")

In [ ]:
# Validamos los cambios


## Transformación de datos
Actividad 2 - Aplicar filtros y transformaciones para crear una tabla de ventas que muestre solo los productos con alto rendimiento.

In [ ]:
df_ventas.head(2)

In [ ]:
# Transformación:
# Calculamos el valor de cada venta = precio * cantidad
# Agregamos una nueva columnas
df_ventas["valor_venta"] = df_ventas["precio"] * df_ventas["cantidad"]

In [ ]:
# Visualizamos
df_ventas.head(2)

In [ ]:
# Agregación:
# Agrupar valor_venta por producto
# Sumamos las ventas dentro de cada grupo
tabla_ventas = (df_ventas.groupby("producto", as_index=False)["valor_venta"].sum())
tabla_ventas.head()

In [ ]:
# Usemos describe para analizar algunos valores del dataframe resultante
tabla_ventas.describe()

In [ ]:
# Agregamos un plot para visualizar graficamente la distribución
import seaborn as sns
import matplotlib.pyplot as plt

sns.histplot(tabla_ventas["valor_venta"], bins=30, kde=True, edgecolor="black")
plt.title("Distribución de Valor de Venta con Curva de Densidad")
plt.xlabel("Valor de venta")
plt.ylabel("Frecuencia")
plt.show()

In [ ]:
# Calcular el percentil 75 (umbral de alto rendimiento)
umbral_ventas = tabla_ventas["valor_venta"].quantile(0.75)
umbral_ventas

In [ ]:
# Identificar productos que superan el umbral
# productos_top = tabla_ventas.query("valor_venta >= @umbral")
productos_top = tabla_ventas[tabla_ventas["valor_venta"] >= umbral_ventas]
productos_top

## Agregación
Actividad 3 - Resumir las ventas por categoría de producto y analizar los ingresos generados.

In [ ]:
df_ventas.columns

In [ ]:
# Agrupamos por categoria y agregamos la suma
ventas_categoria = df_ventas.groupby("categoria", as_index=False)["valor_venta"].sum()
ventas_categoria.head()

In [ ]:
ventas_categoria

In [ ]:
# Ordenar y mostrar los resultados
ventas_categoria

## Integración de datos
Actividad 4 - Combinar los sets de datos de ventas y marketing para obtener una visión más amplia de las tendencias.

In [ ]:
df_ventas.info()

In [ ]:
df_marketing.info()

In [ ]:
# Vemos la cantidad de productos unicos en ventas
df_ventas["producto"].nunique()

In [ ]:
# Vemos la cantidad de productos unicos en marketing
df_marketing["producto"].nunique()

In [ ]:
# Listamos los registros de ventas ordenados por producto ascendiente
df_ventas.sort_values(by="producto", ascending=True )

In [ ]:
# Listamos los registros de marketing ordenados por producto ascendiente
df_marketing.sort_values(by=["producto","canal"])

In [ ]:
# Para evitar duplicados, habria que agrupar marketing antes de hacer el merge (a costa de perder la fecha)
df_marketing_avg = df_marketing.groupby(["producto"], as_index=False)["costo"].mean()
df_marketing_avg

In [ ]:
# Unir los datasets ventas y marketing por columna 'producto'
# Aqui tienen dos caminos, unir ventas con df_marketing que tiene duplicados
ventas_marketing = pd.merge(df_ventas, df_marketing, on="producto", how="left")
# O, unir ventar con df_marketing_avg que no tiene duplicados, pero no tiene fechas
# ventas_marketing = pd.merge(df_ventas, df_marketing_avg, on="producto", how="inner")

In [ ]:
# Verificar coincidencias (muestra)
ventas_marketing.sort_values(by=["producto", "precio"], ascending=True)

# Nota1: si unieron ventas con marketing, obtienen 8994 registros
# Nota2: si unieron ventas con marketing_avg obtienen 2998 registros

### Analizamos el impacto de la campaña

In [ ]:
# Nos quedamos solo con los registros de ventas durante la campaña de marketing
# Nota: si trabajan con el datafrane marketing pueden aplicar el filtro,
# si trabajan con dataframe marketing_avg, al no tener fecha, no apliquen el filtro.

ganancias_marketing = ventas_marketing.query("fecha_venta > fecha_inicio & fecha_venta < fecha_fin")

# Seleccionamos solo los campos que nos interesan: "producto", "valor_venta", "costo"
ganancias_marketing = ganancias_marketing[["producto", "valor_venta", "costo"]]
ganancias_marketing.sort_values(by="producto", ascending=True)

In [ ]:
# Agrupamos por producto y hacemos una agregación por "valor de venta" y "costo"
ganancias_marketing = ganancias_marketing.groupby("producto")[["valor_venta","costo"]].sum()

In [ ]:
# Generamos una nueva columna con la diferencia entre lo vendido y lo invertido en la campaña
ganancias_marketing["ganancia_neta"] = ganancias_marketing['valor_venta']-ganancias_marketing['costo']
ganancias_marketing=ganancias_marketing.sort_values(by="ganancia_neta", ascending=False)

In [ ]:
# Crear un gráfico de barras apilado para comparar valor_venta y costo por producto
ganancias_marketing[["valor_venta", "costo"]].plot(kind='bar', stacked=True, figsize=(12, 6))
plt.title("Comparación de Valor de Venta y Costo por Producto")
plt.xlabel("Producto")
plt.ylabel("Monto")
plt.xticks(rotation=90)
plt.tight_layout()
plt.show()

# Solución al problema de los duplicados

In [ ]:
# Hacemos el merge entre marketing y ventas por columna producto
marketing_ventas = pd.merge(df_marketing, df_ventas, on="producto", how="inner")
marketing_ventas.sort_values(by="producto", ascending=True)

In [ ]:
# filtramos solo aquellos productos que se hayan vendido dentro de la campaña
marketing_ventas_pp1 = marketing_ventas.query("fecha_venta > fecha_inicio & fecha_venta < fecha_fin")

In [ ]:
# Calculamos el total_ventas por producto haciendo valor_venta * cantidad
marketing_ventas_pp2 = marketing_ventas_pp1.copy() # para eliminar la advertencia
marketing_ventas_pp2["total_venta"] = (
    marketing_ventas_pp2["valor_venta"] * marketing_ventas_pp2["cantidad"]
)


In [ ]:
# Agrupamos el dataframe por producto usando .agg
# Agregar total_venta como el promedio, elimina el impacto de los datos duplicados
# Y seguimos agegando costo como la suma de lo invertido en cada canal
ganancias_marketing = marketing_ventas_pp2.groupby("producto").agg(
    {
        "total_venta": "mean",
        "costo": "sum",
    }
)

In [ ]:
# A partir de venta_total y costo, calculamos la ganancia obtenida
ganancias_marketing["ganancia"]= ganancias_marketing["total_venta"]-ganancias_marketing["costo"]
ganancias_marketing = ganancias_marketing.sort_values(by="ganancia", ascending=False)
ganancias_marketing.head(3)

In [ ]:
# Crear un gráfico de barras apilado para comparar valor_venta y costo por producto
ganancias_marketing[["total_venta", "costo"]].plot(kind='bar', stacked=True, figsize=(12, 6))
plt.title("Comparación de Valor de Venta y Costo por Producto")
plt.xlabel("Producto")
plt.ylabel("Monto")
plt.xticks(rotation=90)
plt.tight_layout()
plt.show()

Ahora en el gráfico los costos de campañas de marketing sin más signficativos.